Remove GET:audiences endpoint and getAudiences selector

ankitrox opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #8487 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

techanvil commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTiles.stories.js:
```diff
@@ -124,7 +124,7 @@ export default {
 				registry
 					.dispatch( MODULES_ANALYTICS_4 )
-					.receiveGetAudiences( { audiences: audiencesFixture } );
+					.setAvailableAudiences( { audiences: audiencesFixture } );
```
This should provide an array of objects that represent the cached audiences, rather than the underlying fixtures.

I.e. they should have the shape as described in the AC for https://github.com/google/site-kit-wp/issues/8486 (with the exception of the "All Users" `displayName` which should be mapped to "All visitors" as part of this issue).

techanvil commented on includes/Modules/Analytics_4.php:
```diff
@@ -2177,9 +2160,10 @@ private function filter_analytics_allow_tracking_disabled( $allowed ) {
 				$audience_item = array(
 					'name'        => $audience->getName(),
-					'displayName' => $audience->getDisplayName(),
+					'displayName' => ( 'All Users' === $display_name ) ? __( 'All Visitors', 'google-site-kit' ) : $display_name,
```
Please note the value for "All visitors" should have a lower-case initialled "visitors", and it should also _not_ be translated (the value it's replacing is not translated and is always created in English):

```suggestion
					'displayName' => ( 'All Users' === $display_name ) ? 'All visitors' : $display_name,
```

techanvil commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTiles.js:
```diff
@@ -60,9 +60,12 @@ export default function AudienceTiles( { Widget } ) {
-		select( MODULES_ANALYTICS_4 ).getAudiences()
-	);
+	const audiences = useSelect( ( select ) => {
+		const audienceData =
+			select( MODULES_ANALYTICS_4 ).getAvailableAudiences();
+
+		return Object.values( audienceData );
```
This call to `Object.value()` should no longer be needed:

```suggestion
		return select( MODULES_ANALYTICS_4 ).getAvailableAudiences();
```

techanvil commented on assets/js/modules/analytics-4/components/dashboard/AudienceSegmentation/AudienceTiles.stories.js:
```diff
@@ -124,7 +124,7 @@ export default {
 				registry
 					.dispatch( MODULES_ANALYTICS_4 )
-					.receiveGetAudiences( { audiences: audiencesFixture } );
+					.setAvailableAudiences( mappedCachedAudiences );
```
@ankitrox sorry if this suggestion was a bit unclear - it was for illustrative purposes, you'll still need to create `mappedCachedAudiences` (feel free to use a different name for it too).

ankitrox replied:
@techanvil Thank you. I spotted this that I committed it by mistake, I changed it to `audiencesFixture` which should work.

techanvil replied:
Thanks, @ankitrox - but, this still needs an update. As per my previous comment, the audience fixtures represent the audiences as returned from the GA4 API.

This is not representative of the mapped versions of the audience objects that will be cached and therefore stored in the `availableAudiences` setting.

For reference here's the [audiences fixture](https://github.com/google/site-kit-wp/blob/cc4f31ffd6ed48ca1a4c4464d8b4551b76500274/assets/js/modules/analytics-4/datastore/__fixtures__/audiences.json#L41), as you can see it contains a set of [`Audience` resources](https://developers.google.com/analytics/devguides/config/admin/v1/rest/v1alpha/properties.audiences) which have the following shape:

```ts
{
  "name": string,
  "displayName": string,
  "description": string,
  "membershipDurationDays": integer,
  "adsPersonalizationEnabled": boolean,
  "eventTrigger": {
    object ([AudienceEventTrigger](https://developers.google.com/analytics/devguides/config/admin/v1/rest/v1alpha/properties.audiences#AudienceEventTrigger))
  },
  "exclusionDurationMode": enum ([AudienceExclusionDurationMode](https://developers.google.com/analytics/devguides/config/admin/v1/rest/v1alpha/properties.audiences#AudienceExclusionDurationMode)),
  "filterClauses": [
    {
      object ([AudienceFilterClause](https://developers.google.com/analytics/devguides/config/admin/v1/rest/v1alpha/properties.audiences#AudienceFilterClause))
    }
  ]
}
```

As discussed in https://github.com/google/site-kit-wp/issues/8486, these should be mapped to a set of objects with the following shape which are then cached in the `availableAudiences` setting:

```ts
{
  "name": string,
  "displayName": string,
  "description": string,
  "audienceType": string, // `"DEFAULT_AUDIENCE"`, `"SITE_KIT_AUDIENCE"` or `"USER_AUDIENCE"`.
  "audienceSlug"`: string, // `"all-users"`, `"purchasers"`, `"new-visitors"`, `"returning-visitors"`, or an empty string.
}
```

So, the audiences fixture needs mapping to objects with the above shape to pass to `setAvailableAudiences()`. It might be worth creating a new fixture, say `available-audiences.json`, rather than mapping these in code for the tests etc.

techanvil commented on assets/js/modules/analytics-4/datastore/__fixtures__/available-audiences.json:
```diff
@@ -0,0 +1,38 @@
+        "description": "Description",
+        "displayName": "Test Audience",
+        "audienceType": "USER_AUDIENCE",
+        "audienceSlug": "test-users"
```
Hey @ankitrox just a quick heads up - please note the `audienceSlug` is set to an empty string for user-defined audiences, as can be seen here:

https://github.com/google/site-kit-wp/blob/109d7505a594b2d8db4cb499c14e4abb98b92a8a/includes/Modules/Analytics_4.php#L2244-L2245

```suggestion
        "audienceSlug": ""
```

techanvil commented on assets/js/modules/analytics-4/datastore/__fixtures__/available-audiences.json:
```diff
@@ -0,0 +1,38 @@
+    {
+        "name": "properties/12345/audiences/1",
+        "description": "All users",
+        "displayName": "All Users",
```
@ankitrox, also the `displayName` here should be `"All visitors"` to reflect how it will be mapped as per the related change in this PR:

```suggestion
        "displayName": "All visitors",
```

techanvil replied:
@ankitrox, also the `displayName` here should be `"All visitors"` to reflect how it will be mapped as per the related change in this PR:

```suggestion
        "displayName": "All visitors",
```

techanvil reviewed (changes requested):
Hi @ankitrox, thanks for making these changes.

There are just a couple of aspects that were missed from previous comments:

From https://github.com/google/site-kit-wp/pull/8606#discussion_r1580895591:

> Therefore, please can you update this condition as follows, and please also add a bit of test coverage for this aspect

As per the quoted point, please can you add a bit of test coverage for the changes included via the `getAvailableAudiences()` resolver? I.e. that it should sync the audiences when the setting is `null`.

From https://github.com/google/site-kit-wp/pull/8606#pullrequestreview-2024817608:

> Also, the QA Brief will need an update as it's currently focused on the change to `createAudience()` which will be removed. This should instead reflect the change to `getAvailableAudiences()` (i.e. the sync should be triggered when the value is initially `null`), and it's also worth including a QAB point to give the `AudienceTiles` component a quick check in Storybook to verify it's unchanged.

As per the above, please can you updated the QA Brief accordingly. Thanks!

ankitrox commented:
> Hi @ankitrox, thanks for making these changes.
> 
> There are just a couple of aspects that were missed from previous comments:
> 
> From [#8606 (comment)](https://github.com/google/site-kit-wp/pull/8606#discussion_r1580895591):
> 
> > Therefore, please can you update this condition as follows, and please also add a bit of test coverage for this aspect
> 
> As per the quoted point, please can you add a bit of test coverage for the changes included via the `getAvailableAudiences()` resolver? I.e. that it should sync the audiences when the setting is `null`.
> 
> From [#8606 (review)](https://github.com/google/site-kit-wp/pull/8606#pullrequestreview-2024817608):
> 
> > Also, the QA Brief will need an update as it's currently focused on the change to `createAudience()` which will be removed. This should instead reflect the change to `getAvailableAudiences()` (i.e. the sync should be triggered when the value is initially `null`), and it's also worth including a QAB point to give the `AudienceTiles` component a quick check in Storybook to verify it's unchanged.
> 
> As per the above, please can you updated the QA Brief accordingly. Thanks!

Thank you @techanvil . 

I have addressed following two points.

1. Added the test for `getAvailableAudiences` resolver that it should (and should not call) `sync-audiences` endpoint. 
2. Added the QAB for `getAvailableAudiences` that it should call `sync-audiences` endpoint.
3. Added QAB for comparing `AudienceTiles` component in storybook.

techanvil reviewed (commented):
Thanks @ankitrox, this LGTM!

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/audiences.js:
```diff
@@ -144,42 +114,32 @@ const baseReducer = ( state, { type } ) => {
-	 */
-	getAudiences( state ) {
-		return state.audiences;
+		return registry.select( MODULES_ANALYTICS_4 ).getAvailableAudiences();
```
@techanvil @ankitrox resolvers don't have a return value :) This line can simply be removed.

techanvil replied:
Thanks for spotting that, I totally missed it at review time. I'm including a fix in my current PR.

aaemnnosttv commented on assets/js/modules/analytics-4/datastore/audiences.test.js:
```diff
@@ -159,19 +157,75 @@ describe( 'modules/analytics-4 audiences', () => {
+			];
+
+			it( 'should not sync cached audiences when the availableAudiences setting is not null', () => {
+				fetchMock.postOnce( syncAvailableAudiencesEndpoint, {
+					body: availableAudiences,
+					status: 200,
+				} );
```
Using fetchMock here shouldn't be necessary since the expectation is for it not to be fetched. @ankitrox

techanvil replied:
Thanks here too, @aaemnnosttv, I am removing this call in my current PR too.
